openapi-generator-annotations/lib/src/openapi_generator_annotations_base.dart

```diff
@@ -148,6 +168,234 @@ class Openapi {
+  /// [path] The path of the OAS spec from the root of the [bucket] without the
+  ///   leading /.
+  const AwsRemoteSpec({
+    String? accessKeyId,
+    String? secretAccessKey,
+    required this.region,
+    required this.bucket,
+    required super.path,
+    this.now,
+  })  : _secretAccessKey = secretAccessKey,
+        _accessKeyId = accessKeyId;
+
+  /// The url of the OAS spec within S3.
+  @override
+  Uri get url => Uri.https('$bucket.s3.$region.amazonaws.com', '/$path');
+
+  /// The [Authentication] header content.
+  ///
+  /// This builds the Authorization header content format used by AWS.
+  ///
+  /// Multiple calls will result in the same content unless [toHeaderMap] has been
+  /// called in between.
+  String get authHeaderContent {
+    if (_secretAccessKey == null || _accessKeyId == null) {
+      return '';
+    }
+    // https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html#RESTAuthenticationExamples
+    String toSign = [
+      'GET',
+      '',
+      '',
+      now,
+      '/$bucket/$path',
+    ].join('\n');
+
+    final utf8AKey = utf8.encode(_secretAccessKey!);
+    final utf8ToSign = utf8.encode(toSign);
+
+    final signature =
+        base64Encode(Hmac(sha1, utf8AKey).convert(utf8ToSign).bytes);
+    return 'AWS $_accessKeyId:$signature';
+  }
+
+  /// Load the user credentials from the environment.
+  AwsRemoteSpec loadCredentials() => AwsRemoteSpec(
+        region: region,
+        bucket: bucket,
+        path: path,
+        accessKeyId: Platform.environment['AWS_ACCESS_KEY_ID'],
+        secretAccessKey: Platform.environment['AWS_SECRET_ACCESS_KEY'],
+        now: DateTime.now(),
+      );
+
+  /// Builds the headers map for the authenticated request.
+  ///
+  /// This will return an empty map when when the [_accessKeyId] or [_secretAccessKey]
+  /// are empty after attempting to fetch them from the environment.
+  Map<String, String> toHeaderMap() {
+    AwsRemoteSpec specAuth = this;
+    if (_accessKeyId == null || _secretAccessKey == null || now == null) {
+      specAuth = loadCredentials();
+      if (specAuth.accessKeyId == null ||
+          specAuth.secretAccessKey == null ||
+          now == null) {
+        // TODO: This should probably throw or assert
+        return {};
+      }
+    }
+
+    return {
+      'Authorization': authHeaderContent,
+      'x-amz-date': now!.toIso8601String(),
+    };
+  }
+
+  Map<String, String> toJsonMap() {
+    return {
+      if (_accessKeyId != null) 'accessKeyId': _accessKeyId!,
+      if (_secretAccessKey != null) 'secretAccessKey': _secretAccessKey!,
+      'path': path,
+      'bucket': bucket,
+      'region': region,
+      ...super.toJsonMap(),
+    };
+  }
+
+  AwsRemoteSpec.fromMap(Map<String, dynamic> map)
+      : bucket = map['bucket'],
+        region = map['region'],
+        _accessKeyId = map['accessKeyId'],
+        _secretAccessKey = map['secretAccessKey'],
+        now = DateTime.now(),
```

Comment:
I'm likely going to set this as null to trigger toHeaderMap to reload the creds as `now` needs to be correct within up to 15 minutes of the request time